pkg/kudoctl/util/repo/repo_framework.go

```diff
@@ -0,0 +1,220 @@
+package repo
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/yaml"
+)
+
+// FrameworkRepository represents a framework repository
+type FrameworkRepository struct {
+	Config *RepositoryConfiguration
+	Client HTTPClient
+}
+
+// FrameworkBundle contains parsed files from the framework bundle
+type FrameworkBundle struct {
+	Framework        *v1alpha1.Framework
+	FrameworkVersion *v1alpha1.FrameworkVersion
+	Instance         *v1alpha1.Instance
+}
+
+const (
+	frameworkFileName = "-framework.yaml"
+	versionFileName   = "-frameworkversion.yaml"
+	instanceFileName  = "-instance.yaml"
+)
+
+// NewFrameworkRepository constructs FrameworkRepository
+func NewFrameworkRepository(cfg *RepositoryConfiguration) (*FrameworkRepository, error) {
+	u, err := url.Parse(cfg.URL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid chart URL format: %s", cfg.URL)
+	}
+
+	client, err := NewHTTPClient()
+	if err != nil {
+		return nil, fmt.Errorf("could not construct protocol handler for: %s error: %v", u.Scheme, err)
+	}
+
+	return &FrameworkRepository{
+		Config: cfg,
+		Client: *client,
+	}, nil
+}
+
+// DownloadIndexFile fetches the index file from a repository.
+func (r *FrameworkRepository) DownloadIndexFile() (*IndexFile, error) {
+	var indexURL string
+	parsedURL, err := url.Parse(r.Config.URL)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing config url")
+	}
+	parsedURL.Path = strings.TrimSuffix(parsedURL.Path, "/") + "/index.yaml"
+
+	indexURL = parsedURL.String()
+
+	resp, err := r.Client.Get(indexURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting index url")
+	}
+
+	indexBytes, err := ioutil.ReadAll(resp)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading index response")
+	}
+
+	indexFile, err := parseIndexFile(indexBytes)
+	return indexFile, err
+}
+
+// DownloadBundle downloads the tgz file from the given repo
+func (r *FrameworkRepository) DownloadBundle(bundleName string) (*FrameworkBundle, error) {
+	var fileURL string
+	parsedURL, err := url.Parse(r.Config.URL)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing config url")
+	}
+	parsedURL.Path = parsedURL.Path + "/" + bundleName + ".tgz"
```

Comment:
What's the point of the index file if the bundle can be determined by a name?  I had expected an index file to be required here to get the location of the bundle.

Reply:
Exactly. We need to update the KEP accordingly but this is also been captured/discussed in Slack, e.g. here https://kubernetes.slack.com/archives/CG3HTFCMV/p1559574030076700 

To put it into other words, having the index file `urls` field used to point to the download direction for the tgz.

What we would need here is traversing through the `urls` field, e.g. trying the first url, if it fails then trying the next and so on until the download succeeds. Wdyt?